src/report/report.c

```diff
@@ -0,0 +1,231 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <getopt.h>
+
+#include "alloc-util.h"
+#include "build.h"
+#include "dirent-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "sd-varlink.h"
+#include "string-util.h"
+#include "sd-event.h"
+#include "time-util.h"
+
+#define RUN_SYSTEMD_METRICS_PATH "/run/systemd/metrics/"
+#define MAX_CONCURRENT_METRICS_SOCKETS 20
+#define TIMEOUT_USEC (30*USEC_PER_SEC) /* 30 seconds */
+
+typedef struct Context {
+        int n_ref;
+} Context;
+
+static int metrics_on_query_reply(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+        assert(link);
+
+        Context* context = ASSERT_PTR(userdata);
+
+        if (error_id) {
+                bool disconnect = streq(error_id, SD_VARLINK_ERROR_DISCONNECTED);
+                bool timout = streq(error_id, SD_VARLINK_ERROR_TIMEOUT);
+                if (disconnect)
+                        log_info("Varlink disconnected.");
```

Comment:
nit: ideally, you should log which exact connection disconnected/timeout/errors.